api/utils/accessio/downloader/git/downloader.go

```diff
@@ -0,0 +1,112 @@
+package git
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"sync"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/storage"
+	"github.com/go-git/go-git/v5/storage/memory"
+
+	techgit "ocm.software/ocm/api/tech/git"
+	"ocm.software/ocm/api/utils/accessio/downloader"
+)
+
+const localRemoteName = "origin"
+
+type CloseableDownloader interface {
+	downloader.Downloader
+	Close() error
+}
+
+// Downloader simply uses the default HTTP client to download the contents of a URL.
+type Downloader struct {
+	cloneOpts *git.CloneOptions
+
+	matching *regexp.Regexp
+
+	mu      sync.Mutex
+	buf     *bytes.Buffer
+	storage storage.Storer
+}
+
+var _ downloader.Downloader = (*Downloader)(nil)
+
+func NewDownloader(url string, ref string, path string, auth techgit.AuthMethod) CloseableDownloader {
+	refName := plumbing.ReferenceName(ref)
+	return &Downloader{
+		cloneOpts: &git.CloneOptions{
+			Auth:          auth,
+			URL:           url,
+			RemoteName:    localRemoteName,
+			ReferenceName: refName,
+			SingleBranch:  true,
+			Tags:          git.NoTags,
+			Depth:         0,
+		},
+		matching: regexp.MustCompile(path),
+		buf:      bytes.NewBuffer(make([]byte, 0, 4096)),
+		storage:  memory.NewStorage(),
+	}
+}
+
+func (d *Downloader) Download(w io.WriterAt) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	ctx := context.Background()
+
+	// no support for git archive yet, so we need to clone the repository in bare mode
+	repo, err := git.CloneContext(ctx, d.storage, nil, d.cloneOpts)
+	if err != nil {
+		return fmt.Errorf("failed to clone repository %s: %w", d.cloneOpts.URL, err)
+	}
+
+	trees, err := repo.TreeObjects()
+	if err != nil {
+		return fmt.Errorf("failed to get tree objects: %w", err)
+	}
+
+	if err := trees.ForEach(func(t *object.Tree) error {
+		return t.Files().ForEach(d.copyFileToBuffer)
+	}); err != nil {
+		return fmt.Errorf("failed to iterate over trees: %w", err)
+	}
+
+	defer d.buf.Reset()
+	if _, err := w.WriteAt(d.buf.Bytes(), 0); err != nil {
+		return fmt.Errorf("failed to write blobs: %w", err)
+	}
+
+	return nil
+}
+
+func (d *Downloader) copyFileToBuffer(file *object.File) error {
+	if !d.matching.MatchString(file.Name) {
+		return nil
+	}
+
+	reader, err := file.Reader()
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(d.buf, reader)
+	return errors.Join(err, reader.Close())
+}
+
```

Comment:
Hmm, what kind of blob are you creating here, you just concatenate the content of the files? How do you want to unpack this again to a file structure. This should be an archive with the appropriate file structure.

Reply:
Currently Ioriented myself on the other downloaders, namely s3 and http which only download single files, not entire folders. If that desire is there I can certianly add a tar/zipping on client side. I thought that wouldnt be necessary though right now

Reply:
The current logic only downloads the file matching to the name specified in the downloader

Reply:
This downloader is not used by the access method anymore. We can remove it or leave it in for convenience